### PR TITLE
Enable torch build with SLEEF on ARM by default

### PR DESCRIPTION
**Pull request by aditew01**
**Scope:** Enable PyTorch build with SLEEF on Arm by default. Enable codegen kernels compilation with SLEEF on ARM platform.

Enabling the build with SLEEF by default and setting `AT_BUILD_ARM_VEC256_WITH_SLEEF` as the default for Arm  improves performance for some models. I have benchmarked several networks on `Neoverse-V1` using `torch.compile` with the `inductor` backend.
On models  like `hf_Bert_Large` , `hf_GPT_fast`, we're seeing a **~1.2x speedup** (with 16 threads).

The below results are run with `Batch_Size=1` and `Cores=8, 16`

Screenshot 2024-08-27 at 17 04 23

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @malfet @milpuz01 @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

**Comment by aditew01** [comment]
cc: @malfet

**Review by malfet — APPROVED** [review]

**Comment by aditew01** [comment]
@malfet I see one more job triggered yesterday for this PR. https://github.com/pytorch/pytorch/actions/runs/10420725278 
There was one failure trace: `error: HTTP Error 409: Public access is not permitted on this storage account.`  which maybe related to this: https://github.com/pytorch/pytorch/issues/134073 ?

**Comment by cfRod** [comment]
label "ciflow/linux-aarch64"

**Comment by aditew01** [comment]
@malfet can this be merged now ?

File: torch/_inductor/cpu_vec_isa.py [hunk 1]
```diff
@@ -317,8 +315,6 @@ def _check_and_append_supported_isa(
 @functools.lru_cache(None)
 def valid_vec_isa_list() -> List[VecISA]:
     isa_list: List[VecISA] = []
-    if sys.platform == "darwin" and platform.processor() == "arm":
-        isa_list.append(VecNEON())
```
Comment (malfet) [inline torch/_inductor/cpu_vec_isa.py:321]: Why are you removing vectorization support for MacOS?

**Review by malfet — CHANGES_REQUESTED** [review]
Your PR disables Vectorization on MacOS, isn't it?

File: torch/_inductor/cpu_vec_isa.py [hunk 2]
```diff
@@ -337,9 +333,10 @@ def valid_vec_isa_list() -> List[VecISA]:
                         if re.search(r"[\^ ]+vxe[\$ ]+", group):
                             isa_list.append(VecZVECTOR())
                             break
+
     elif arch == "ppc64le":
         isa_list.append(VecVSX())
-    elif arch == "aarch64":
+    elif arch in ["aarch64", "arm64"]:
```
Comment (malfet) [inline torch/_inductor/cpu_vec_isa.py:339]:
```suggestion
    elif arch == "aarch64":
```

Comment (aditew01) [reply torch/_inductor/cpu_vec_isa.py:339]: "arm64" is for keeping vectorization for MacOS

File: torch/_inductor/cpu_vec_isa.py (hunk 1 above)
Comment (aditew01) [reply torch/_inductor/cpu_vec_isa.py]: I assumed adding platform.machine as "arm64" would enable vec for macOS, but apparently that does not resolve the macOS story. I'll revert the commit with this change. Thanks!

**Review by malfet — APPROVED** [review]
Looks good to me, thank you for the updates

**Review by kimishpatel — APPROVED** [review]

**Review by kimishpatel — APPROVED** [review]

**Comment by aditew01** [comment]
> https://github.com/pytorch/pytorch/blob/e8ad5085739ef6d680f22818f2fd53c897a425de/CMakeLists.txt#L308
> 
> Does this boolean do anything after your proposed change? Should it be removed?

I think this is a good suggestion given the flags will now be activated by default. Thanks for pointing it out @robert-hardwick

**Comment by maajidkhann** [comment]
We have also tested "**SVE + Inductor flow**" from @aditew01 **PR:** (https://github.com/pytorch/pytorch/pull/134672) without sleef and with sleef and observed consistent performance improvements with sleef build.

This change will further enhance default performance on ARM CPU's.

**The below results are from torchbench on a 32 core Graviton 3 EC2 Instance:**

image

changes LGTM!

**Comment by aditew01** [comment]
@malfet a naive question, do we re-trigger the mergebot?

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
> @malfet a naive question, do we re-trigger the mergebot?

Yes, you should be able to, I was waiting for Android fix before issuing another command.

**Comment by blapie** [comment]
Hello! SLEEF maintainer speaking here. I have a few questions regarding this PR. 
1. Why does it have to be disabled on Android? Is there a plan to enable it?
2. What is compared in the benchmarks? Is it comparing against calls to standard scalar implementations (e.g. libc)? When SLEEF is enabled, is it calling to Neon or SVE routines?
3. I suppose only the high accuracy routines are used, since they have the typical GNU ABI names?
4. Is there room for relaxing accuracy and use the more competitive 3.5ULP (standard accuracy in glibc libmvec)?

**Pull request closed without merging**